WIP: add glossary with some basic DVC terms

algomaster99 opened the pull request:
This PR just initializes the glossary. Suggestions for more terms and concepts, to be included here, are more than welcome!

Fixes #424

shcheklein commented on static/docs/user-guide/glossary.md:
```diff
@@ -0,0 +1,30 @@
+## Pipeline
+
+In our documentation, it has also been referred to as a computation graph. It is
+an ordered set of some executable commands, each of them stored in a Dvcfile. To
```
DVC-file

jorgeorpinel replied:
**Pipeline** is already explained in https://dvc.org/doc/understanding-dvc/what-is-dvc, https://dvc.org/doc/get-started/pipeline, and https://dvc.org/doc/commands-reference/pipeline

Maybe search the docs for these terms first like I mentioned in https://github.com/iterative/dvc.org/issues/424#issuecomment-500640694

shcheklein commented on static/docs/user-guide/glossary.md:
```diff
@@ -0,0 +1,30 @@
+In our documentation, it has also been referred to as a computation graph. It is
+an ordered set of some executable commands, each of them stored in a Dvcfile. To
+get familiar with its usage, please refer to this
+[tutorial](/doc/get-started/example-pipeline#example-pipelines).
```
better link to get-started or NLP example in get started

shcheklein commented on static/docs/user-guide/glossary.md:
```diff
@@ -0,0 +1,30 @@
+get familiar with its usage, please refer to this
+[tutorial](/doc/get-started/example-pipeline#example-pipelines).
+
+## DAG
```
it's the same as previous?

algomaster99 replied:
@shcheklein it seemed like similar when I read about DAG from some resources online.

jorgeorpinel replied:
Definitely requires disambiguation. TBHI think  this glossary is not a simple task, please take your time trying to unwind these concepts from existing docs before suggesting how to explain them and where to link from the glossary (:

jorgeorpinel replied:
@algomaster99 did you have a chance to check the suggested groups of terms for the glossary I listed in the description of #424 ? In there "Pipeline" and "DAG" are actually separate but related entries.

jorgeorpinel commented:
Hello. I updated the description in #424 with a suggested list of glossary entries. Please take a look. Maybe include those in this initial PR? Thanks

algomaster99 commented:
@jorgeorpinel I will look into it as soon as possible. Thanks :)

algomaster99 commented:
I have added a tooltip component which will show the meanings of DVC terminologies on hovering.

### Demo
![tooltip](https://user-images.githubusercontent.com/35191225/59689165-c6867f00-91fc-11e9-8b74-b405c022bc86.gif)

You can see the working above or try them out in `version.md` file by hovering on top of `workspace` and `cache`.
I have added the pronunciation too since it was looking cool :sweat_smile:  

### Improvements
- [ ] As one can see the explanation of cache is hiding below the sidebar. I tried adjusting the `z-index` but it doesn't work. One solution could be to guess the screen size and make the tooltip toggle in opposite direction to the present state. 
- [ ] On mobile, we will have to activate it by clicking for that I have showed a symbol `✝`. Can anyone tell me how can we go about the implementation?

algomaster99 commented:
@shcheklein @jorgeorpinel Please have a look :)

shcheklein commented on static/docs/commands-reference/version.md:
```diff
@@ -14,14 +14,14 @@ usage: dvc version [-h] [-q | -v]
+| `Python version`                            | Version of the Python being used for the project in which DVC is initialized                                                                                           |
+| `Platform`                                  | Information about the operating system of the machine                                                                                                                  |
+| [`Binary`](#output-of-binary)               | Shows whether the package is installed from a binary release or source                                                                                                 |
+| `Cache`                                     | [Type of links](/doc/user-guide/large-dataset-optimization#file-link-types-for-the-dvc-cache) supported between the DVC workspace and the <abbr>cache</abbr> directory |
```
I'm not sure I like the idea of using <abbr> </abbr> for each term we need to highlight, but have no better ideas from the top of my head ... @jorgeorpinel anything comes to your mind?

algomaster99 replied:
@shcheklein I don't exactly get what you are saying. You don't want to use `abbr` tags and let the code automatically detect the keywords?

jorgeorpinel replied:
It would be ideal if tooltips could be added automatically by the docs engine. Being careful not to interfere with any existing automatic parsing like the `dvc cmd ref` links.

algomaster99 replied:
@jorgeorpinel Okay, I will try to figure out.

algomaster99 replied:
@jorgeorpinel I think it will be better if we manually indicate which terms we need to highlight using `abbr` tags since automatic detection will require checking each word in a markdown file and checking if it exists in `glossary.json`.

jorgeorpinel replied:
I agree this is a complicated feature, lets discuss with Ivan (:

shcheklein replied:
Yeah, don't see a way to make it automatic. Let's just keep on a writer discretion to put the <abbr> tag when it makes sense. Like with the DVC-file link we discussed with @jorgeorpinel it probably does not make sense to put it all around the document. Just a few places? 

We can also show the terms on the left side. And show tooltip when you hover.

shcheklein commented on src/Documentation/glossary.json:
```diff
@@ -0,0 +1,18 @@
+    {
+      "name": "Cache directory",
+      "match": ["cache"],
+      "enun": "/kaʃ dɪˈrɛkt(ə)ri/",
```
That would be too much I think in our case, to include enuns.

algomaster99 replied:
@shcheklein Okay, I'll remove it. Just thought it will look nice :smile:

shcheklein commented on src/Documentation/glossary.json:
```diff
@@ -0,0 +1,18 @@
+{
```
make is a JS file, so that we can pass through prettier

algomaster99 replied:
@shcheklein The pre-commit hook takes care of that. We can just tweak the `npm run prettier-js` command to make it work for json files as well.
Something like this: `prettier --write '{pages,src}/**/*.{js,jsx,json}'`

shcheklein commented on src/Documentation/glossary.json:
```diff
@@ -0,0 +1,18 @@
+      "name": "Workspace directory",
+      "match": ["workspace"],
+      "enun": "/ˈwəːkspeɪs dɪˈrɛkt(ə)ri/",
+      "desc": "Also abbreviated as workspace, it is the root directory of a project where DVC is initialized by running `dvc init` command. Therefore, this directory will contain a `.dvc` directory as well."
```
do we render `dvc` and links? Should we use markdown and ReactMarkdown to render these small snippets?

algomaster99 replied:
@shcheklein Yeah it would be better to render it as markdown.

algomaster99 commented:
@shcheklein I think we can delete `glossary.md` and generate it using the json file instead.

jorgeorpinel commented on src/Documentation/glossary.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "Glossary",
+  "desc": "This guide is aimed to familiarize the users with definitions to relevant DVC concepts and terminologies which are frequently used.",
```
Is is posisble to wrap JSON files to 80 chars? Otherwise a JS file may be preferable after all.

shcheklein replied:
agreed w @jorgeorpinel - we should run it by prettier

jorgeorpinel commented on src/Documentation/sidebar.json:
```diff
@@ -48,6 +48,7 @@
     "name": "User Guide",
     "folder": "/static/docs/user-guide",
     "files": [
+      "glossary.md",
```
The glossary page should probably be the last in this group.

shcheklein replied:
agreed with @jorgeorpinel , let's move it all the way down

jorgeorpinel commented on static/docs/user-guide/glossary.md:
```diff
@@ -0,0 +1,30 @@
+# Glossary
```
Could this page be generated automatically by the engine instead of having a separate .md file for it? The long descriptions would need to be saved in the glossary.json/js file though...

algomaster99 replied:
@jorgeorpinel Even I was thinking about the same.

algomaster99 commented:
@jorgeorpinel @shcheklein 
Here the changes I have made till now:
![tooltip](https://user-images.githubusercontent.com/35191225/60005998-3da09500-968d-11e9-9a4d-89ae8aee7da3.gif)

- The issue with `z-index` has been resolved by controlling the direction it pops in.
- The `enun` (pronunciation) has been removed.
- The content is rendered as markdown in the tooltip.

algomaster99 commented:
Now two things need to be tackled:
- Generate the markdown file using `glossary.json`. Where this process should happen? And how should we register the created file in `sidebar.json`?
- Enabling click, in addition, to hovering since this will make our tooltip component render on mobile views too. 
- Also, is json fine or should I switch to `glossary.js`?

algomaster99 commented:
> There should be a section or a few sections called Basic Concepts in the user guide that is properly on a user level language explains some missing concepts

I don't get this exactly. How will we go about doing this? @shcheklein

shcheklein commented:
@algomaster99 an example would a DVC-file term and this link - https://dvc.org/doc/user-guide/dvc-file-format . We don't need a separate glossary.md just to have a link to this doc. This doc + floating tips with a summary and a link should be enough.

algomaster99 commented:
@shcheklein Okay. I have one more question which is that where all should we add the `abbr` tags to show the tooltip?

algomaster99 commented:
@jorgeorpinel So we start writing short notes about things listed below right?
> - Workspace, DVC project - related: SCM repository, working directory
> - DVC meta files and directories (in '.dvc/' dir) - related: DVC config file, DVC cache
> - DVC cache, cache, "the cache" - related: file checksum, file link, innode, remote cache
> - DVC-file - related: stage file, .dvc file extension, Dvcfile file name, orphan stage file
> - DVC remote, remote cache - related: remote location (see external deps/outs guides), cache
> - Pipeline, data pipeline - related: command chain, pipeline stage
> - Stage, pipeline stage - related: stage file, DVC-file
> - DAG, computational graph, command chain - related: pipeline
> - Metric?
> - DVC workflow? - related: workflow, Git workflow, experiment workflow

jorgeorpinel commented:
@algomaster99 re: `where all should we add the <abbr> tags to show the tooltip?`

Ivan said `Let's just keep on a writer discretion to put the tag when it makes sense...
it probably does not make sense to put it all around the document. Just a few places?`

So just adding `<abbr>` around terms that exist in glossary.json does the trick? I suggest to open a separate issue to review all the terms in all the docs and decide doc by doc which occurrences to <abbr> (many of them are currently linked for example so it requires using criteria on a case-by-case basis) AFTER this initial glossary.json is finished and merged. So this PR can focus on finishing the initial glossary and finish the engine updates.

jorgeorpinel commented:
BTW Ivan also said

- [ ] `We can also show the terms on the left side. And show tooltip when you hover.`

So that would be another feature to finish for this PR if you and him confirm about it. Also...

- [ ] Can we remove the symbols and just have a dotted underline for <abbr> terms? It's easier to finger touch and better looking. Similar to this:
  ![image](https://user-images.githubusercontent.com/1477535/60208909-15e13680-981f-11e9-8d7e-dd27393ff106.png)

algomaster99 commented:
@jorgeorpinel I can explain the working more clearly.
So adding `abbr` tags like this: `<abbr>cache</abbr>` will get detected and this will traverse the object in `glossary.js` file for an attribute called `match`. `match` is an array of keywords which match a specific description.
For example:
```
{
      name: 'Cache directory',
      match: ['cache'],
      desc:
        'DVC cache is a hidden storage which is found at `.dvc/cache`. This ' +
        'storage is used to manage different versions of files which are ' +
        'under DVC control. For more information on cache, please refer to ' +
        'the this [guide](/doc/commands-reference/config#cache).'
}
```
Now why `match` is an array?
This will help us highlighting terms like `cache`, `cache directory` and still have the same description for them.

algomaster99 commented:
@jorgeorpinel Where on the left side exactly? I don't get it.

algomaster99 commented:
The underline looks something like this:
![Screenshot from 2019-06-27 03-37-19](https://user-images.githubusercontent.com/35191225/60218406-e6691300-988c-11e9-84f1-3847b1422caf.png)

shcheklein commented:
Sorry for the confusion, I meant the right side 🤦‍♂ Just beneath the edit and chat buttons. No need to all of this at once. Let's not overcomplicate it.

 Underlying looks great, @algomaster99

shcheklein commented on src/Documentation/glossary.js:
```diff
@@ -0,0 +1,25 @@
+      name: 'Workspace directory',
+      match: ['workspace'],
+      desc:
+        'Also abbreviated as workspace, it is the root directory of a ' +
```
It not necessarily the root directory. It all the files that are visible to the user in the project directory. I would check if there is a good explanation for Git.

jorgeorpinel replied:
Agree. The name can just be `Workspace` as it refers to the entire DVC project or repo from what I understand. So both these terms should be mentioned as synonyms. Remember that the glossary helps us either disambiguate similar terms or confirm synonyms.

jorgeorpinel replied:
Suggested desc: "The **workspace** or DVC **project** consists of all the files and directories under the location where DVC was initialized with `dvc init`. Its root directory will contain a `.dvc/` subdirectory with the special [DVC Files and Directories] (/doc/user-guide/dvc-files-and-directories)". Typically, the workspace also contains a Git **repository**, which intersects with the DVC project, but they're not exactly the same.

shcheklein replied:
@jorgeorpinel sounds a little bit complicated :) Too much information. The question is - what do we want user to understand there? Most likely that workspace == his/her project files/directories. Then we can mention the usually it a git repo, and .dvc, etc.

jorgeorpinel replied:
How about:

The **workspace** contains all of your DVC **project** files and directories. It's typically also a Git **repository**. See also `dvc init` (should link to cmd ref).

shcheklein replied:
Yep, looks almost perfect!

I would probably do `DVC project` -> `project`. 

 ... files, directories and DVC-files

link to init + a link to DVC Files

shcheklein commented on src/Documentation/glossary.js:
```diff
@@ -0,0 +1,25 @@
+      name: 'Cache directory',
+      match: ['cache'],
+      desc:
+        'DVC cache is a hidden storage which is found at `.dvc/cache`. This ' +
```
it's usually `.dvc/cache` - not always

jorgeorpinel replied:
Use `name: 'DVC Cache`, and you can probably copy a `desc` from https://dvc.org/doc/user-guide/dvc-files-and-directories and/or https://dvc.org/doc/commands-reference/cache

> BTW we'll have to review if we need to unify al these definitions throughout docs... But that's for a future issue i.e. #461

shcheklein replied:
please address this comment

shcheklein commented:
we need to fix this issue (show it down):

<img width="1326" alt="Screen Shot 2019-06-26 at 4 22 53 PM" src="https://user-images.githubusercontent.com/3659196/60222263-85602180-9831-11e9-99b7-7904501305b5.png">

jorgeorpinel commented:
> Note: I extracted the stuff about adding `abbr` tags and possibly the sidebar terms to #461

algomaster99 commented:
@shcheklein I am facing issues to position it downwards when scrolled. Somehow, I cannot capture the amount scrolled using `window.scrollY`. It always shows `0` no matter how much I scroll.

shcheklein commented:
@algomaster99 check the `doc.js` code. It's using `react-scroller`. Probably you should try to call it?

algomaster99 commented:
@shcheklein this package is for animating scrolls, right. I only need the scroll position.

jorgeorpinel commented:
I'm not familiar enough with React yet to help with this issue, sorry!

I did want to mention that I added `<abbr>` around the term `data artifact` in the changes of PR #464 so please add the term to `glossary.js`. BTW can the `match` property be an array @algomaster99? In order to include the plural form of the term, in this case `data artifacts` (but adding an `s` to the end doesn't always work so I think an array is needed).

algomaster99 commented:
@jorgeorpinel it is already an array

algomaster99 commented:
@jorgeorpinel What should be the description of this term?

algomaster99 commented:
@shcheklein @jorgeorpinel 
How do we detect if it is a mobile view or a desktop view? [This package](https://www.npmjs.com/package/react-device-detect) does it but I am sure our project uses something similar to it.
I need this information to render the tooltip properly in mobile view as well.
The current solution could be to pass states like this:
```
state = {
    hover: false,
    timeout: null,
    width: 400,
    margin: -70,
    pointMargin: -15,
    mobileWidth: xx,
    mobileMargin: xx,
  }
```
What should I follow?

jorgeorpinel commented:
> What should be the description of this term? (`data artifact`)

I would list it after `DVC meta files and directories` in the glossary order, and its a simple description: "Any **data** file, intermediate result, or final product such as a ML model file, that is under DVC control. Refer to [Data and Model Files Versioning](/doc/use-cases/data-and-model-files-versioning)."

Not sure about the React Q, sorry. I need to get up to date on React and on our docs engine soon (:

algomaster99 commented:
@jorgeorpinel will make the changes once #464 is merged

jorgeorpinel commented:
No need to wait. Not sure how long that PR will take to be ready. Who ever finishes first avoids the possible conflicts 😉

algomaster99 commented:
@jorgeorpinel 
![Screenshot from 2019-07-12 01-28-17](https://user-images.githubusercontent.com/35191225/61081378-aab17a00-a444-11e9-9dd0-f052f60d8351.png)

P.S. I have deleted the `data artifact` term from this doc. It was just there for demonstration.

shcheklein commented:
@algomaster99 any updates on the JS code here? Can we ask some help on SO, do more experiments, or something to figure this out?

algomaster99 commented:
@shcheklein I was working on the mobile view for this now which is almost ready but I'm still stuck on getting the scroll position so that I can toggle the position of the tooltip between top and bottom.

jorgeorpinel commented:
Maybe extract the work for mobile to another issue/pull @algomaster99 ? Also please note a huge PR just got merged, please rebase your local repo from upstream, hopefully there won't be conflicts. Thanks!

algomaster99 commented:
@jorgeorpinel okay, we can keep the mobile view for another issue.
@shcheklein why isn't `window.scrollY` working in the project? :confused:

shcheklein commented:
@algomaster99 can we do some research for the scroll position? Ask on SO for example?

algomaster99 commented:
@shcheklein What is SO?

shcheklein commented:
@algomaster99 StackOverflow

algomaster99 commented:
@shcheklein There are already a lot of questions there related to this. They have suggested `window.scrollY` or `window.pageYOffset` to get it. Doesn't work for us. But still, I'll ask the question and put a link here.

algomaster99 commented:
@shcheklein Let's wait for [this](https://stackoverflow.com/questions/57070385/cant-get-scroll-position-using-window-scrolly-or-window-pageyoffset) to resolve.

algomaster99 commented:
@jorgeorpinel @shcheklein 
![tooltip](https://user-images.githubusercontent.com/35191225/61489951-849e5380-a9c9-11e9-9728-9ef6d3978265.gif)

jorgeorpinel commented:
Great! So what was the correct answer in your SO question? Let's upvote it :P 

This PR is now mergeable then? (Have you rebased from upstream yet?) Thanks!

p.s. I created a review app for this PR if you want to do a quick sanity check: https://dvc-org-pr-431.herokuapp.com/doc – Looks good to me except that when a term in `<abbr>` doesn't yet exist in the glossary.js file, the underline or tooltip shouldn't render at all:

![image](https://user-images.githubusercontent.com/1477535/61500123-18026500-a987-11e9-800f-c78e24251a3e.png)

shcheklein commented on src/Tooltip/index.js:
```diff
@@ -0,0 +1,228 @@
+    const maxWidth = markdownBody.offsetLeft + markdownBody.clientWidth
+    const container = document.getElementsByClassName('tooltip-container')[0]
+    const tooltipWidth = container.offsetLeft + this.state.width
+    const vertical = tooltipHeight > 80 ? 'top' : 'bottom'
```
can we take 80 from actual container, not hardcode it?

algomaster99 commented:
@jorgeorpinel @shcheklein 
Yeah, you can upvote it. There is only one answer :)
Please review as well. I have made some changes.

algomaster99 commented:
@shcheklein It is currently not designed for the mobile view so it is breaking a little - like it can go off-screen, etc.

jorgeorpinel commented:
@algomaster99 

> currently not designed for the mobile view so it is breaking a little

Did you ever open a separate issue to address the mobile view? Or add it to the existing future issue #461 please.

shcheklein commented:
@algomaster99 one more comment left to address (specify that `.dvc/cache` is the default location) + can we disable it on mobile if it breaks?

algomaster99 commented:
@jorgeorpinel @shcheklein I have made the changes. You can look at it :)
